Sources/ComposableArchitecture/Reducer.swift

```diff
@@ -290,39 +290,162 @@ public struct Reducer<State, Action, Environment> {
+  ///     child actions when the child domain is non-`nil`.
+  ///
+  ///         IfLetStore(
+  ///           self.parentStore.scope(state: { $0.child }, action: { .child($0) }
+  ///         ) { childStore in
+  ///           // This destination only appears when child state is non-`nil`
+  ///           WithViewStore(childStore) { childViewStore in
+  ///             // So this action can only be sent when child state is non-`nil`
+  ///             Button("Child Action") { childViewStore.send(.action) }
+  ///           }
+  ///           ...
+  ///         }
   ///
   /// - See also: `IfLetStore`, a SwiftUI helper for transforming a store on optional state into a
   ///   store on non-optional state.
   /// - See also: `Store.ifLet`, a UIKit helper for doing imperative work with a store on optional
   ///   state.
-  public func optional(_ file: StaticString = #file, _ line: UInt = #line) -> Reducer<
+  ///
+  /// - Parameter breakpointOnNil: Raises `SIGTRAP` signal when an action is sent to the reducer
+  ///   but state is `nil`. This is generally considered a logic error, as a child reducer cannot
+  ///   process a child action for unavailable child state.
+  /// - Returns: A reducer that works on optional state.
+  public func optional(
+    breakpointOnNil: Bool = true,
+    _ file: StaticString = #file,
+    _ line: UInt = #line
+  ) -> Reducer<
     State?, Action, Environment
   > {
     .init { state, action, environment in
       guard state != nil else {
-        assertionFailure(
-          """
-          "\(debugCaseOutput(action))" was received by an optional reducer when its state was \
-          "nil". This can happen for a few reasons:
-
-          * The optional reducer was combined with or run from another reducer that set \
-          "\(State.self)" to "nil" before the optional reducer ran. Combine or run optional \
-          reducers before reducers that can set their state to "nil". This ensures that optional \
-          reducers can handle their actions while their state is still non-"nil".
-
-          * An active effect emitted this action while state was "nil". Make sure that effects for \
-          this optional reducer are canceled when optional state is set to "nil".
-
-          * This action was sent to the store while state was "nil". Make sure that actions for \
-          this reducer can only be sent to a view store when state is non-"nil". In SwiftUI \
-          applications, use "IfLetStore".
-          """,
-          file: file,
-          line: line
-        )
+        #if DEBUG
+          if breakpointOnNil {
+            fputs(
+              """
+              ---
+              Warning: Reducer.optional@\(file):\(line))
+
+              "\(debugCaseOutput(action))" was received by an optional reducer when its state was \
+              "nil". This is generally considered an application logic error, and can happen for a \
```

Comment:
Qualified this with "generally" everywhere now and note that it is reasonable to ignore the result of some effects in a bullet point below.

Reply:
@stephencelis In my experience this almost exclusively happens because of erratic NavigationLink binding behavior. The action in the binding gets called multiple times, for mysterious reasons. Maybe it's worthwhile to list it as one of the sub points and an Apple bug?

Reply:
I can confirm that NavigationLinks are weird in my CA apps too, and my I also blame SwiftUI 😅.

But @ferologics, doesn't something as simple as sending an action from `onDisappear` count?